### DISCUSSION TITLE
Listening on same IPv4 and IPv6 port?

### DISCUSSION THREAD

**Discussion by jcorporation** [cesanta/mongoose#2305] [Ask questions, get answers] [answered]
I am using the http-server from the example and run it with `./example -l http://[::]:8080`. Should mongoose listen also on IPv4 0.0.0.0:8080 in this case?

Plattform is: Linux t14 6.2.0-25-generic #25-Ubuntu SMP PREEMPT_DYNAMIC Fri Jun 16 17:05:07 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

```
cat /proc/sys/net/ipv6/bindv6only
0

wget http://127.0.0.1:8080 -O - 
--2023-07-18 23:16:14--  http://127.0.0.1:8080/
Connecting to 127.0.0.1:8080... failed: Connection refused.
```

**Comment by cpq** [comment]
@jcorporation why should it, if you have specified the ipv6  listening address ?

**Reply by jcorporation** [reply]
Older versions of mongoose had this behavior.

As I understand the socket code: not setting IPV6_V6ONLY enables ipv4 and ipv6 connections. 

I looked into the mongoose source: If I remove this code from mg_open_listener() function:

```
#if defined(IPV6_V6ONLY)
    } else if (c->loc.is_ip6 &&
               (rc = setsockopt(fd, IPPROTO_IPV6, IPV6_V6ONLY, (char *) &on,
                                sizeof(on))) != 0) {
      // See #2089. Allow to bind v4 and v6 sockets on the same port
      MG_ERROR(("setsockopt(IPV6_V6ONLY): %d", MG_SOCK_ERR(rc)));
#endif
```

Mongoose accepts IPv6 and IPv4 connections. It would be great if you could add an compile time option to enable/disable this behavior.

It was changed with #2089.

**Reply by comburent** [reply]
The description in #2089 seems contradictory: it fixes "not listening to the same port on both ipv4 and ipv6" by not listening to both? It's unclear what exactly was the original issue there

**Reply by comburent** [reply]
> @jcorporation why should it, if you have specified the ipv6 listening address ?

Per <https://www.rfc-editor.org/rfc/rfc3493#section-5.3>
>  This socket option restricts AF_INET6 sockets to IPv6 communications
   only.  As stated in section <3.7 Compatibility with IPv4 Nodes>,
   AF_INET6 sockets may be used for both IPv4 and IPv6 communications.

Reading the RFC, the `IPV6_V6ONLY` is oriented for specific scenarios, not as a default.

Further  references:
- <https://www.kernel.org/doc/html/latest/networking/ip-sysctl.html#proc-sys-net-ipv6-variables>

**Comment by jcorporation** [comment]
Should I maintain my own patch or is there any chance that this behavior will change? A compile time option would be perfect for me.

**Comment by cpq** [comment, marked as answer]
@jcorporation 

What if we change :

```c
#if defined(IPV6_V6ONLY)
```

to 

```c
#if MG_IPV6_V6ONLY
```

And define MG_IPV6_V6ONLY as 0 by default, would that work ?

**Reply by jcorporation** [reply]
This sounds good!
